File: dataforge-core/src/main/java/hep/dataforge/workspace/BasicWorkspace.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.workspace;

import hep.dataforge.cache.CachePlugin;
import hep.dataforge.context.Context;
import hep.dataforge.data.Data;
import hep.dataforge.data.DataNode;
import hep.dataforge.data.DataTree;
import hep.dataforge.meta.Meta;
import hep.dataforge.workspace.tasks.Task;
import hep.dataforge.workspace.tasks.TaskModel;
import org.jetbrains.annotations.NotNull;

/**
 * A basic data caching workspace
 *
 * @author Alexander Nozik
 */
public class BasicWorkspace extends AbstractWorkspace {

    private DataTree.Builder<Object> data = DataTree.builder();
    private transient CachePlugin cache;

    @NotNull
    public static Builder builder() {
        return new Builder();
    }

    @Override
    public DataNode<Object> getData() {
        return data.build();
    }

    protected boolean cacheEnabled() {
        return true;
    }

    public DataNode<Object> runTask(TaskModel model) {
        //Cache result if cache is available and caching is not blocked
        if (cacheEnabled() && model.meta().getBoolean("cache.enabled", true)) {
            Task<Object> task = getTask(model.getName());
            return getCache().cacheNode(model.getName(), model.getIdentity(), task.run(model));
        } else {
            return super.runTask(model);
        }
    }

    protected synchronized CachePlugin getCache() {
        if (cache == null || cache.getContext() != this.getContext()) {
            cache = getContext().optFeature(CachePlugin.class).orElseGet(() -> {
                CachePlugin pl = new CachePlugin();
                getContext().pluginManager().load(pl);
                return pl;
            });
        }
        return cache;
    }

    @Override
    public void clean() {
        getLogger().info("Cleaning up cache...");
        invalidateCache();
    }

    public void invalidateCache() {
        if (cacheEnabled()) {
            getCache().invalidate();
        }
    }


    public static class Builder implements Workspace.Builder {

        BasicWorkspace w = new BasicWorkspace();

        @Override
        public Builder self() {
            return this;
        }

        @Override
        public Context getContext() {
            return w.getContext();
        }

        @Override
        public Builder setContext(Context ctx) {
            w.setContext(ctx);
            return self();
        }

        private DataTree.Builder getData() {
            return w.data;
        }

        @Override
        @SuppressWarnings("unchecked")
        public Builder loadData(String as, Data<?> data) {
            if (w.getData().optNode(as).isPresent()) {
                getLogger().warn("Overriding non-empty data during workspace data fill");
            }
            getData().putData(as, data);
            return self();
        }

        @Override
        @SuppressWarnings("unchecked")
        public Builder loadData(String as, DataNode<?> datanode) {
            if (as == null || as.isEmpty()) {
                if (!w.data.isEmpty()) {
                    getLogger().warn("Overriding non-empty root data node during workspace construction");
                }
                w.data = new DataTree.Builder(datanode);
            } else {
                getData().putNode(as, datanode);
            }
            return self();
        }

        @Override
        public Builder loadTask(Task task) {
            w.tasks.put(task.getName(), task);
            return self();
        }

        @Override
        public Builder target(String name, Meta meta) {
            w.targets.put(name, meta);
            return self();
        }

        @Override
        public Workspace build() {
            return w;
        }

    }

}

File: dataforge-core/src/main/kotlin/hep/dataforge/workspace/BasicWorkspace.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.workspace

import hep.dataforge.cache.CachePlugin
import hep.dataforge.context.Context
import hep.dataforge.context.Global
import hep.dataforge.data.Data
import hep.dataforge.data.DataNode
import hep.dataforge.data.DataNodeEditor
import hep.dataforge.data.DataTree
import hep.dataforge.kodex.optional
import hep.dataforge.meta.Meta
import hep.dataforge.workspace.tasks.Task
import hep.dataforge.workspace.tasks.TaskModel

/**
 * A basic data caching workspace
 *
 * @author Alexander Nozik
 */
class BasicWorkspace private constructor(
        context: Context,
        override val data: DataNode<*>,
        override val taskMap: Map<String, Task<*>>,
        override val targetMap: Map<String, Meta>) : AbstractWorkspace(context) {

    private val cache: CachePlugin by lazy {
        context.opt(CachePlugin::class.java).optional.orElseGet {
            val pl = CachePlugin(Meta.empty())
            context.pluginManager.load(pl)
            pl
        }
    }

    private fun cacheEnabled(): Boolean {
        return context.getBoolean("immutable.enabled", true)
    }

    override fun runTask(model: TaskModel): DataNode<*> {
        //Cache result if immutable is available and caching is not blocked
        return if (cacheEnabled() && model.meta.getBoolean("immutable.enabled", true)) {
            cache.cacheNode(model.name, model.toMeta(), super.runTask(model))
        } else {
            super.runTask(model)
        }
    }

    override fun clean() {
        logger.info("Cleaning up cache...")
        invalidateCache()
    }

    private fun invalidateCache() {
        if (cacheEnabled()) {
            cache.invalidate()
        }
    }


    class Builder(override var context: Context = Global) : Workspace.Builder {
        private var data: DataNodeEditor<Any> = DataTree.edit(Any::class.java).apply { name = "data" }

        private val taskMap: MutableMap<String, Task<*>> = HashMap<String, Task<*>>()
        private val targetMap: MutableMap<String, Meta> = HashMap()

        //internal var workspace = BasicWorkspace()

        override fun self(): Builder {
            return this
        }

        override fun data(key: String, data: Data<out Any>): Builder {
//            if (this.data.optNode(key) != null) {
//                logger.warn("Overriding non-empty data during workspace data fill")
//            }
            this.data.putData(key, data)
            return self()
        }

        override fun data(key: String?, dataNode: DataNode<out Any>): Builder {
            if (key == null || key.isEmpty()) {
                if (!data.isEmpty) {
                    logger.warn("Overriding non-empty root data node during workspace construction")
                }
                data = dataNode.edit() as DataNodeEditor<Any>
            } else {
                data.putNode(key, dataNode)
            }
            return self()
        }

        override fun task(task: Task<*>): Builder {
            taskMap[task.name] = task
            return self()
        }

        override fun target(name: String, meta: Meta): Builder {
            targetMap[name] = meta
            return self()
        }

        override fun build(): Workspace {
            context.pluginManager.stream(true)
                    .flatMap { plugin -> plugin.provideAll(Task.TASK_TARGET, Task::class.java) }
                    .forEach { taskMap.putIfAbsent(it.name, it) }
            return BasicWorkspace(context, data.build(), taskMap, targetMap)
        }

    }

    companion object {

        fun builder(): Builder {
            return Builder()
        }
    }

}
